File: Julia simulations/ridging test.R
library("np")
library(tidyverse)
library(DEoptim)
data <- read_csv("try data.csv")
source("auxiliary functions.R")

# OLS
summary(lm(y ~ x, data = data))
# 2SLS
summary(AER::ivreg(y ~ x|z, data = data))
# WMDF
wmd( x = data$x , y = data$y , z = data$z, exo = NULL, intercept = FALSE )

pred <- lcll( dependent = data$x, explanatory = data$z, bandwidth = 0.001, type = "ll", kertype = "gaussian", loo = FALSE )  # Ridging has been implemented
np_pred <- npreg(bws = 0.001, txdat = data$z, tydat = data$x, regtype = "ll")$mean
np_pred - pred


# With the Epanechnikov kernel. Does not seem to be the same thing
pred <- lcll( dependent = data$x, explanatory = data$z, bandwidth = 0.1, type = "ll", kertype = "epanechnikov", loo = FALSE )  # Ridging has been implemented
np_pred <- npreg(bws = 0.1, txdat = data$z, tydat = data$x, regtype = "ll", ckertype = "epanechnikov")$mean
np_pred - pred

# Check leave-one-out CV criterion function
loocv(dep = data$x, expl = data$z, bw = 0.001, type = "ll")
loocv(dep = data$x, expl = data$z, bw = 1, type = "ll")


# Bandwidth computation
np_bw <- npregbw( xdat = data$z, ydat = data$x, regtype = "ll")
my_bw <- opti_h(dep = data$x, expl = data$z, type = "ll" )
np_bw
my_bw

# Things match pretty well between np and my functon opti_h, although the optimal bandwidth is different
loocv( dep = data$x, expl = data$z, bw = my_bw$par, type = "ll" ) # loo CV criterion at the optimal "opti_h" bandwidth
loocv( dep = data$x, expl = data$z, bw = np_bw$bw, type = "ll" ) # loo CV criterion at the optimal np bandwidth
lcll( dependent = data$x, explanatory = data$z, bandwidth = my_bw$par, type = "ll", kertype = "gaussian", loo = FALSE )  # Ridging has been implemented
lcll( dependent = data$x, explanatory = data$z, bandwidth = np_bw$bw, type = "ll", kertype = "gaussian", loo = FALSE )  # Ridging has been implemented



# On a bigger data set
n <- 1000
data <- dgp(n = n, g = "A&L")
# Bandwidth computation
np_bw <- npregbw( xdat = data$z1, ydat = data$x, regtype = "ll")
my_bw <- opti_h(dep = data$x, expl = data$z1, type = "ll" )
np_bw
my_bw  

loocv( dep = data$x, expl = data$z, bw = my_bw$par, type = "ll" ) # loo CV criterion at the optimal "opti_h" bandwidth
loocv( dep = data$x, expl = data$z, bw = np_bw$bw, type = "ll" ) # loo CV criterion at the optimal np bandwidth
head(lcll( dependent = data$x, explanatory = data$z, bandwidth = my_bw$par, type = "ll", kertype = "gaussian", loo = FALSE ))  # Ridging has been implemented
head(lcll( dependent = data$x, explanatory = data$z, bandwidth = np_bw$bw, type = "ll", kertype = "gaussian", loo = FALSE ) ) # Ridging has been implemented

# Bandwidth computation with AIC criterion (Hurvic, Simonov and Tsai)
np_bw <- npregbw( xdat = data$z, ydat = data$x, regtype = "ll", bwmethod = "cv.aic")
my_bw <- optimr::optimr(par = c(1), fn = hurvic_AIC, method = "Nelder-Mead", explanatory = data$z, dependent = data$x, type = "ll", loo = FALSE)
# They match!! 
np_bw
my_bw

# Try the iv version now. Works bof so far (find negative bandwidth sometimes)
min_AIC_iv(expl = data$y, endo = data$x, instru = data$z1, type = "ll", loo = FALSE )
  
# Try the loo CV for betahat. It works!
loo_opti(x = data$x, y = data$y, z = data$z1, regtype = "ll")
np_2SLS(Y = data$y, endo = data$x, exo = NULL, Z = data$z1, type = "ll", intercept = FALSE)
  
  

File: Julia simulations/ridging test.jl
# This script tests the kernel esitmators "faits maison"
using CSV
using DataFrames
using KernelEstimator

include("auxiliary functions.jl")
include("Estimators.jl")

data = CSV.read("F:\\Kernel 2SLS\\Julia simulations\\try data.csv", DataFrame)   # Laptop computer
#data = CSV.read("G:\\Kernel 2SLS\\Julia simulations\\try data.csv", DataFrame)   # Desktop computer
# OLS
ols(data.x, data.y; intercept = "true")
# 2SLS
tsls(data.x, data.y, data.z; intercept = "true")
# WMDF
wmdf(data.x, data.y, data.z; intercept = "false")
# Kernel regression
pred = lcll( data.x, data.z; bw = 0.001)
print(pred[15, :])
print(pred[21, :])
print(pred[25, :])
print(pred[31, :])

np_pred = npr(data.z, data.x, xeval=data.z, reg=locallinear, kernel=gaussiankernel, h = 0.001)

show(stdout, "text/plain", pred)
show(stdout, "text/plain", np_pred)

# Look at the leave-one-out CV criterion function
loocv(data.x, data.z, 0.001, "lc" )
loocv(data.x, data.z, 1, "lc" )
loocv(data.x, data.z, 9539536, "lc" )

# Look at the optimal bandwidth using leave-one-out CV
op_h = opti_bandwidth(data.x, data.z, "lc" )
loocv(data.x, data.z, op_h, "lc" )
pred = lcll( data.x, data.z; bw = op_h, type = "lc" )  # don't mind the "lc", right now only the local linear is coded
pred

# Look at the information criterion from Hurvic Simonoff and Tsai (1997)
iv_AIC(data.y, data.x, data.z; bw = 0.1, type = "lc")
min_AIC_iv(data.y, data.x, data.z; lower = 0.01, upper = 50.0, type = "lc") #, loo = FALSE )

# Look at the optimal bandwidh with a leave-one-out MSE criterion function
kernel_TV = loo_opti(data.x, data.y, data.z, regtype = "ll")
bw_TV = kernel_TV.Bandwidth[1]

# Estimate with the CV criterion function
myest_bw(data.x, data.y, data.z; regtype = "lc")


#=
include("auxiliary function.jl")
 n = 1000
 mu = [0.0, 0]
 sigma = [1.0, 0.5, 0.5, 1.0]
 sigma = reshape(sigma, 2, 2)
 data = dgp(n, 1, mu, sigma)
 AIC_h = hurvic_AIC(data.y, data.x, bw = 0.1)
 AIC_iv_h = iv_AIC(data.y, data.x, data.z; bw = 0.1, type = "ll")
=#
